File: internal/cashflow/forecast.go
// Package cashflow calculates dated scenarios without changing the actual ledger.
package cashflow

import (
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"math"
	"sort"
	"time"
)

// Amount uses integer minor units and a JSON string, including for currencies
// whose smallest unit is not a cent.
type Amount int64

func (a Amount) MarshalJSON() ([]byte, error) { return json.Marshal(fmt.Sprint(int64(a))) }
func (a *Amount) UnmarshalJSON(b []byte) error {
	var s string
	if err := json.Unmarshal(b, &s); err != nil {
		return fmt.Errorf("amount must be a minor-unit string")
	}
	var n int64
	if _, err := fmt.Sscan(s, &n); err != nil || fmt.Sprint(n) != s {
		return fmt.Errorf("invalid integer amount")
	}
	*a = Amount(n)
	return nil
}

type Account struct {
	Code     string `json:"code"`
	Name     string `json:"name"`
	Kind     string `json:"kind"` // bank or card; card liabilities are negative
	Opening  Amount `json:"opening"`
	Floor    Amount `json:"floor"`
	Reserved bool   `json:"reserved"`
	Evidence string `json:"evidence"`
}
type Event struct {
	ID        string `json:"id"`
	Date      string `json:"date"`
	Name      string `json:"name"`
	Kind      string `json:"kind"` // inflow, outflow, transfer
	Account   string `json:"account"`
	ToAccount string `json:"to_account,omitempty"`
	Arrival   string `json:"arrival,omitempty"`
	Amount    Amount `json:"amount"`
	Status    string `json:"status"` // confirmed, estimated, proposed, actual
	Category  string `json:"category,omitempty"`
	Evidence  string `json:"evidence"`
	Replaces  string `json:"replaces,omitempty"` // explicit expected-event ID; never heuristic matching
}
type Plan struct {
	Version     string    `json:"version"`
	Name        string    `json:"name"`
	Currency    string    `json:"currency"`
	AsOf        string    `json:"as_of"` // opening balances are end-of-day
	Through     string    `json:"through"`
	Accounts    []Account `json:"accounts"`
	Events      []Event   `json:"events"` // dated occurrences, including every recurring occurrence
	Assumptions []string  `json:"assumptions"`
}
type Movement struct {
	Event   Event  `json:"event"`
	Delta   Amount `json:"delta"`
	Balance Amount `json:"balance"`
}
type Day struct {
	Date      string     `json:"date"`
	Account   string     `json:"account"`
	Opening   Amount     `json:"opening"`
	Inflow    Amount     `json:"inflow"`
	Outflow   Amount     `json:"outflow"`
	Closing   Amount     `json:"closing"`
	Floor     Amount     `json:"floor"`
	Shortfall Amount     `json:"shortfall"`
	Movements []Movement `json:"movements"`
}
type Total struct {
	Date        string `json:"date"`
	BankCash    Amount `json:"bank_cash"`
	WorkingCash Amount `json:"working_cash"`
	CardBalance Amount `json:"card_balance"`
	InTransit   Amount `json:"in_transit"`
}
type Low struct {
	Account   string `json:"account"`
	Date      string `json:"date"`
	Balance   Amount `json:"balance"`
	Shortfall Amount `json:"shortfall"`
}
type Variance struct {
	Expected         Event  `json:"expected"`
	Actual           Event  `json:"actual"`
	AmountDifference Amount `json:"amount_difference"`
}
type Result struct {
	Version   string     `json:"version"`
	Digest    string     `json:"digest"`
	Plan      Plan       `json:"plan"`
	Days      []Day      `json:"days"`
	Totals    []Total    `json:"totals"`
	Lows      []Low      `json:"lows"`
	Variances []Variance `json:"variances"`
	Warnings  []string   `json:"warnings"`
}

func date(s string) (time.Time, error) {
	t, e := time.Parse("2006-01-02", s)
	if e != nil || t.Format("2006-01-02") != s {
		return t, fmt.Errorf("invalid date %q", s)
	}
	return t, nil
}
func add(a, b Amount) (Amount, error) {
	if (b > 0 && a > Amount(math.MaxInt64)-b) || (b < 0 && a < Amount(math.MinInt64)-b) {
		return 0, fmt.Errorf("cash amount overflow")
	}
	return a + b, nil
}
func gap(f, b Amount) (Amount, error) {
	if b >= f {
		return 0, nil
	}
	if b == Amount(math.MinInt64) {
		return 0, fmt.Errorf("cash amount overflow")
	}
	return add(f, -b)
}

// Project is deterministic. Opening balances include all events through AsOf.
// Plans are immutable inputs; callers retain input/output digests for comparison.
func Project(p Plan) (Result, error) {
	r := Result{Version: "books.cashflow/v1", Plan: p, Days: []Day{}, Totals: []Total{}, Lows: []Low{}, Variances: []Variance{}, Warnings: []string{}}
	start, err := date(p.AsOf)
	if err != nil {
		return r, err
	}
	end, err := date(p.Through)
	if err != nil {
		return r, err
	}
	if p.Version != "books.cash-plan/v1" || p.Name == "" || len(p.Accounts) == 0 || len(p.Accounts) > 100 || len(p.Events) > 20000 || end.Before(start) || end.Sub(start) > 730*24*time.Hour {
		return r, fmt.Errorf("invalid plan version, name, size or date range")
	}
	accounts := map[string]Account{}
	balances := map[string]Amount{}
	lows := map[string]Low{}
	for _, a := range p.Accounts {
		if a.Code == "" || a.Name == "" || a.Evidence == "" || (a.Kind != "bank" && a.Kind != "card") || a.Floor < 0 || a.Opening == Amount(math.MinInt64) {
			return r, fmt.Errorf("invalid account %q", a.Code)
		}
		if _, ok := accounts[a.Code]; ok {
			return r, fmt.Errorf("duplicate account %q", a.Code)
		}
		accounts[a.Code] = a
		balances[a.Code] = a.Opening
		g := Amount(0)
		if a.Kind == "bank" {
			g, err = gap(a.Floor, a.Opening)
			if err != nil {
				return r, err
			}
		}
		lows[a.Code] = Low{a.Code, p.AsOf, a.Opening, g}
	}
	events := map[string]Event{}
	replaced := map[string]bool{}
	for _, e := range p.Events {
		if _, err = date(e.Date); err != nil {
			return r, err
		}
		a, ok := accounts[e.Account]
		if !ok || e.ID == "" || e.Name == "" || e.Evidence == "" || e.Amount <= 0 || (e.Kind != "inflow" && e.Kind != "outflow" && e.Kind != "transfer") || (e.Status != "confirmed" && e.Status != "estimated" && e.Status != "proposed" && e.Status != "actual") {
			return r, fmt.Errorf("invalid event %q", e.ID)
		}
		if _, ok = events[e.ID]; ok {
			return r, fmt.Errorf("duplicate event %q", e.ID)
		}
		if e.Kind == "transfer" {
			b, ok := accounts[e.ToAccount]
			if !ok || e.ToAccount == e.Account || a.Kind != "bank" || (b.Kind != "bank" && b.Kind != "card") {
				return r, fmt.Errorf("invalid transfer %q", e.ID)
			}
			if _, err = date(e.Arrival); err != nil {
				return r, err
			}
			if e.Arrival < e.Date {
				return r, fmt.Errorf("arrival precedes departure")
			}
		} else if e.ToAccount != "" || e.Arrival != "" {
			return r, fmt.Errorf("non-transfer has a destination")
		}
		events[e.ID] = e
	}
	for _, e := range p.Events {
		if e.Replaces != "" {
			old, ok := events[e.Replaces]
			if !ok || old.Replaces != "" || old.Status == "actual" || e.Status != "actual" || e.Account != old.Account || e.Kind != old.Kind || e.ToAccount != old.ToAccount || replaced[old.ID] {
				return r, fmt.Errorf("invalid actual match %q", e.ID)
			}
			replaced[old.ID] = true
			r.Variances = append(r.Variances, Variance{old, e, e.Amount - old.Amount})
		}
	}
	type leg struct {
		event   Event
		delta   Amount
		transit Amount
	}
	calendar := map[string]map[string][]leg{}
	transit := Amount(0)
	put := func(day, account string, l leg) {
		if calendar[day] == nil {
			calendar[day] = map[string][]leg{}
		}
		calendar[day][account] = append(calendar[day][account], l)
	}
	for _, e := range p.Events {
		if replaced[e.ID] {
			continue
		}
		if e.Date <= p.AsOf && e.Status != "actual" {
			return r, fmt.Errorf("unresolved event %q is on or before opening date; match an actual or explicitly reschedule it", e.ID)
		}
		delta := e.Amount
		if e.Kind != "inflow" {
			delta = -delta
		}
		if e.Date > p.AsOf && e.Date <= p.Through {
			tr := Amount(0)
			if e.Kind == "transfer" {
				tr = e.Amount
			}
			put(e.Date, e.Account, leg{e, delta, tr})
		}
		if e.Kind == "transfer" {
			if e.Date <= p.AsOf && e.Arrival > p.AsOf {
				transit, err = add(transit, e.Amount)
				if err != nil {
					return r, err
				}
			}
			if e.Arrival > p.AsOf && e.Arrival <= p.Through {
				put(e.Arrival, e.ToAccount, leg{e, e.Amount, -e.Amount})
			}
		}
		if e.Status == "proposed" && e.Date <= p.Through && (e.Date > p.AsOf || e.Arrival > p.AsOf) {
			r.Warnings = append(r.Warnings, "Includes proposed transfer/event: "+e.ID)
		}
	}
	for day := start; !day.After(end); day = day.AddDate(0, 0, 1) {
		ds := day.Format("2006-01-02")
		total := Total{Date: ds}
		for _, a := range p.Accounts {
			row := Day{Date: ds, Account: a.Code, Opening: balances[a.Code], Floor: a.Floor, Movements: []Movement{}}
			legs := calendar[ds][a.Code]
			sort.Slice(legs, func(i, j int) bool {
				if legs[i].delta < 0 != (legs[j].delta < 0) {
					return legs[i].delta < 0
				}
				return legs[i].event.ID < legs[j].event.ID
			})
			for _, l := range legs {
				balances[a.Code], err = add(balances[a.Code], l.delta)
				if err != nil {
					return r, err
				}
				transit, err = add(transit, l.transit)
				if err != nil {
					return r, err
				}
				if l.delta > 0 {
					row.Inflow, err = add(row.Inflow, l.delta)
				} else {
					row.Outflow, err = add(row.Outflow, -l.delta)
				}
				if err != nil {
					return r, err
				}
				row.Movements = append(row.Movements, Movement{l.event, l.delta, balances[a.Code]})
			}
			row.Closing = balances[a.Code]
			if a.Kind == "bank" {
				row.Shortfall, err = gap(a.Floor, row.Closing)
				if err != nil {
					return r, err
				}
				total.BankCash, err = add(total.BankCash, row.Closing)
				if err != nil {
					return r, err
				}
				if !a.Reserved {
					total.WorkingCash, err = add(total.WorkingCash, row.Closing)
				}
			} else {
				total.CardBalance, err = add(total.CardBalance, row.Closing)
			}
			if err != nil {
				return r, err
			}
			if row.Closing < lows[a.Code].Balance {
				lows[a.Code] = Low{a.Code, ds, row.Closing, row.Shortfall}
			}
			r.Days = append(r.Days, row)
		}
		total.InTransit = transit
		r.Totals = append(r.Totals, total)
	}
	for _, a := range p.Accounts {
		r.Lows = append(r.Lows, lows[a.Code])
	}
	r.Warnings = append(r.Warnings, "Balances are end-of-day; intraday availability is not established. Planned movements do not execute payments.")
	raw, err := json.Marshal(p)
	if err != nil {
		return r, err
	}
	sum := sha256.Sum256(raw)
	r.Digest = hex.EncodeToString(sum[:])
	return r, nil
}

File: web/src/lib/forecast.ts
// Point-forward interpretation of a books.cash-plan/v1 forecast result. Every
// amount is an integer minor-unit bigint; nothing here uses floating point.
export type ForecastEvent = {
  id: string;
  name: string;
  date: string;
  kind: "inflow" | "outflow" | "transfer";
  arrival?: string;
  account: string;
  to_account?: string;
  amount: string;
  status: "confirmed" | "estimated" | "proposed" | "actual";
  evidence: string;
  category?: string;
  replaces?: string;
};
export type ForecastAccount = {
  code: string;
  name: string;
  kind: "bank" | "card";
  reserved: boolean;
  opening: string;
  floor: string;
  evidence: string;
};
export type ForecastDay = {
  date: string;
  account: string;
  opening: string;
  inflow: string;
  outflow: string;
  closing: string;
  floor: string;
  shortfall: string;
  movements: { event: ForecastEvent; delta: string; balance: string }[];
};
export type Forecast = {
  digest: string;
  scenarios: string[];
  plan: {
    name: string;
    as_of: string;
    through: string;
    currency: string;
    accounts: ForecastAccount[];
    events?: ForecastEvent[];
    assumptions?: string[];
  };
  days: ForecastDay[];
  lows: { account: string; date: string; balance: string; shortfall: string }[];
  warnings: string[];
  variances: {
    expected: ForecastEvent;
    actual: ForecastEvent;
    amount_difference: string;
  }[];
};

export type MonthOutlook = {
  month: string;
  from: string;
  to: string;
  complete: boolean;
  income: bigint;
  spending: bigint;
  leftOver: bigint;
  toReserves: bigint;
  afterReserves: bigint;
  reserves: { code: string; name: string; amount: bigint }[];
  categories: { name: string; amount: bigint }[];
};
export type AccountGap = {
  code: string;
  name: string;
  reserved: boolean;
  floor: bigint;
  firstDate: string;
  lowestDate: string;
  lowest: bigint;
  shortfall: bigint;
  daysShort: number;
};
export type Outlook = {
  asOf: string;
  through: string;
  months: MonthOutlook[];
  complete: MonthOutlook[];
  averageLeftOver?: bigint;
  averageToReserves?: bigint;
  averageAfterReserves?: bigint;
  tightest?: MonthOutlook;
  shortMonths: MonthOutlook[];
  reserveActivity: boolean;
  gaps: AccountGap[];
  counts: Record<ForecastEvent["status"], number> & { total: number };
  accounts: { bank: number; reserved: number; card: number };
};

function lastDay(month: string) {
  const [y, m] = month.split("-").map(Number);
  return `${month}-${String(new Date(Date.UTC(y, m, 0)).getUTCDate()).padStart(2, "0")}`;
}
function later(a: string, b: string) {
  return a > b ? a : b;
}
function earlier(a: string, b: string) {
  return a < b ? a : b;
}
// Rounds half away from zero to the nearest minor unit.
export function average(values: bigint[]): bigint | undefined {
  if (!values.length) return undefined;
  const total = values.reduce((a, b) => a + b, 0n),
    n = BigInt(values.length),
    q = total / n,
    r = total % n;
  const abs = r < 0n ? -r : r;
  return abs * 2n >= n ? q + (total < 0n ? -1n : 1n) : q;
}

/**
 * Groups the scenario's future events by the calendar month money moves.
 * Income is every inflow. Spending is every outflow, including purchases on
 * cards; card payments and transfers are never counted as spending again.
 * Net to reserves is the change in reserved bank balances: transfers in, less
 * transfers out and spending paid from a reserve. Left over less net to
 * reserves is therefore the change in non-reserved money. A month is complete
 * only when the plan covers its first and last day; partial months are shown
 * but excluded from averages.
 */
export function summarizeOutlook(forecast: Forecast): Outlook {
  const { plan } = forecast;
  const accounts = new Map(plan.accounts.map((a) => [a.code, a]));
  const replaced = new Set(
    (plan.events ?? []).flatMap((e) => (e.replaces ? [e.replaces] : [])),
  );
  const horizon = (plan.events ?? []).filter(
    (e) => !replaced.has(e.id) && e.date > plan.as_of && e.date <= plan.through,
  );
  const months = new Map<string, MonthOutlook>();
  const start = new Date(plan.as_of + "T00:00:00Z");
  start.setUTCDate(start.getUTCDate() + 1);
  for (
    let cursor = start.toISOString().slice(0, 7);
    cursor <= plan.through.slice(0, 7);
  ) {
    const from = later(`${cursor}-01`, start.toISOString().slice(0, 10)),
      to = earlier(lastDay(cursor), plan.through);
    months.set(cursor, {
      month: cursor,
      from,
      to,
      complete: from === `${cursor}-01` && to === lastDay(cursor),
      income: 0n,
      spending: 0n,
      leftOver: 0n,
      toReserves: 0n,
      afterReserves: 0n,
      reserves: [],
      categories: [],
    });
    const [y, m] = cursor.split("-").map(Number);
    cursor =
      m === 12 ? `${y + 1}-01` : `${y}-${String(m + 1).padStart(2, "0")}`;
  }
  const reserveTotals = new Map<string, Map<string, bigint>>();
  const categoryTotals = new Map<string, Map<string, bigint>>();
  const bump = (
    store: Map<string, Map<string, bigint>>,
    month: string,
    key: string,
    amount: bigint,
  ) => {
    const inner = store.get(month) ?? new Map<string, bigint>();
    inner.set(key, (inner.get(key) ?? 0n) + amount);
    store.set(month, inner);
  };
  const counts = {
    total: horizon.length,
    confirmed: 0,
    estimated: 0,
    proposed: 0,
    actual: 0,
  };
  for (const e of horizon) {
    counts[e.status]++;
    const month = months.get(e.date.slice(0, 7));
    const source = accounts.get(e.account);
    if (!month || !source) continue;
    const amount = BigInt(e.amount);
    const reserved = (a?: ForecastAccount) => a?.kind === "bank" && a.reserved;
    if (e.kind === "inflow") {
      month.income += amount;
      if (reserved(source)) {
        month.toReserves += amount;
        bump(reserveTotals, month.month, source.code, amount);
      }
    } else if (e.kind === "outflow") {
      month.spending += amount;
      bump(categoryTotals, month.month, e.category || "Uncategorized", amount);
      if (reserved(source)) {
        month.toReserves -= amount;
        bump(reserveTotals, month.month, source.code, -amount);
      }
    } else {
      const destination = accounts.get(e.to_account ?? "");
      if (reserved(source) === reserved(destination)) continue;
      const signed = reserved(destination) ? amount : -amount;
      const code = reserved(destination) ? destination!.code : source.code;
      month.toReserves += signed;
      bump(reserveTotals, month.month, code, signed);
    }
  }
  const list = [...months.values()].map((m) => {
    m.leftOver = m.income - m.spending;
    m.afterReserves = m.leftOver - m.toReserves;
    m.reserves = [...(reserveTotals.get(m.month) ?? new Map())]
      .filter(([, amount]) => amount !== 0n)
      .map(([code, amount]) => ({
        code,
        name: accounts.get(code)?.name ?? code,
        amount,
      }));
    m.categories = [...(categoryTotals.get(m.month) ?? new Map())]
      .map(([name, amount]) => ({ name, amount }))
      .sort((a, b) => (b.amount > a.amount ? 1 : b.amount < a.amount ? -1 : 0));
    return m;
  });
  const complete = list.filter((m) => m.complete);
  const gaps: AccountGap[] = [];
  for (const account of plan.accounts) {
    if (account.kind !== "bank") continue;
    const short = forecast.days.filter(
      (d) => d.account === account.code && BigInt(d.shortfall) > 0n,
    );
    if (!short.length) continue;
    const low = short.reduce((a, b) =>
      BigInt(b.closing) < BigInt(a.closing) ? b : a,
    );
    gaps.push({
      code: account.code,
      name: account.name,
      reserved: account.reserved,
      floor: BigInt(account.floor),
      firstDate: short[0].date,
      lowestDate: low.date,
      lowest: BigInt(low.closing),
      shortfall: BigInt(low.shortfall),
      daysShort: short.length,
    });
  }
  gaps.sort(
    (a, b) =>
      a.firstDate.localeCompare(b.firstDate) || a.name.localeCompare(b.name),
  );
  return {
    asOf: plan.as_of,
    through: plan.through,
    months: list,
    complete,
    averageLeftOver: average(complete.map((m) => m.leftOver)),
    averageToReserves: average(complete.map((m) => m.toReserves)),
    averageAfterReserves: average(complete.map((m) => m.afterReserves)),
    tightest: complete.reduce<MonthOutlook | undefined>(
      (a, m) => (!a || m.leftOver < a.leftOver ? m : a),
      undefined,
    ),
    shortMonths: complete.filter((m) => m.leftOver < 0n),
    reserveActivity: list.some((m) => m.reserves.length > 0),
    gaps,
    counts,
    accounts: {
      bank: plan.accounts.filter((a) => a.kind === "bank").length,
      reserved: plan.accounts.filter((a) => a.kind === "bank" && a.reserved)
        .length,
      card: plan.accounts.filter((a) => a.kind === "card").length,
    },
  };
}

export function monthName(month: string, style: "long" | "short" = "long") {
  return new Date(month + "-01T12:00:00Z").toLocaleDateString("en-US", {
    month: style,
    timeZone: "UTC",
  });
}
export function daysBetween(from: string, to: string) {
  return Math.round(
    (Date.parse(to + "T00:00:00Z") - Date.parse(from + "T00:00:00Z")) /
      86400000,
  );
}

export function scenarioLabel(key: string) {
  const words = key.replace(/[-_]+/g, " ").trim();
  return words.charAt(0).toUpperCase() + words.slice(1);
}
export const monthSpan = (months: MonthOutlook[]) =>
  months.length === 1
    ? monthName(months[0].month, "short")
    : `${monthName(months[0].month, "short")}–${monthName(months.at(-1)!.month, "short")}`;
const list = (names: string[]) =>
  names.length < 3
    ? names.join(" and ")
    : `${names.slice(0, -1).join(", ")} and ${names.at(-1)}`;

// The one-sentence answer shared by the Outlook page and the overview card.
export function verdict(o: Outlook): {
  tone: "covered" | "short" | "none";
  text: string;
} {
  if (!o.complete.length)
    return { tone: "none", text: "This plan doesn’t cover a full month yet." };
  if (o.averageLeftOver! < 0n)
    return {
      tone: "short",
      text: "Planned spending is more than expected income on average.",
    };
  if (o.shortMonths.length)
    return {
      tone: "short",
      text: `Planned spending is more than expected income in ${list(o.shortMonths.map((m) => monthName(m.month)))}.`,
    };
  return {
    tone: "covered",
    text:
      o.complete.length === 1
        ? "Expected income covers planned spending this full month."
        : "Expected income covers planned spending in every full month.",
  };
}
